compiler/rustc_const_eval/src/const_eval/mod.rs

```diff
@@ -90,19 +115,67 @@ fn const_to_valtree_inner<'tcx>(
         // Technically we could allow function pointers (represented as `ty::Instance`), but this is not guaranteed to
         // agree with runtime equality tests.
         ty::FnPtr(_) | ty::RawPtr(_) => None,
-        ty::Ref(..) => unimplemented!("need to use deref_const"),
+
+        ty::Ref(_, inner_ty, _)  => {
+            match inner_ty.kind() {
+                ty::Slice(_) | ty::Str => {
+                    match ecx.try_read_immediate_from_mplace(&place) {
+                        Ok(Some(imm)) => {
+                            let derefd = ecx.deref_operand(&place.into()).expect(&format!("couldnt deref {:?}", imm));
+                            debug!(?derefd);
+
+                            let len = match *imm {
+                                Immediate::ScalarPair(_, b) => {
+                                    let len = b.to_machine_usize(&ecx.tcx.tcx).unwrap();
+                                    len
+                                }
+                                _ => bug!("expected ScalarPair for &[T] or &str"),
+                            };
+                            debug!(?len);
+
+                            let valtree = slice_branches(ecx, &derefd, len);
+                            debug!(?valtree);
+
+                            valtree
+                        }
+                        _ => {
+                            None
+                        }
+                    }
+                }
+                _ => {
+                    let imm = ecx.try_read_immediate_from_mplace(&place).unwrap_or_else(|e| bug!("couldnt read immediate from {:?}, error: {:?}", place, e));
+
+                    match imm {
+                        Some(imm) => {
+                            debug!(?imm);
```

Comment:
`imm` seems to be unused except for debug output -- seems like you could entirely get rid of it?